[action] Integrate Managed Google Play with get_managed_play_store_publishing_rights and create_app_on_managed_play_store

janpio opened the pull request:
### Checklist
- [ ] I've run `bundle exec rspec` from the root directory to see all new and existing tests pass
- [ ] I've followed the _fastlane_ code style and run `bundle exec rubocop -a` to ensure the code style is valid
- [ ] I've read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [ ] I've updated the documentation if necessary.

### Motivation and Context
I was asked to integrate [`fastlane-plugin-managed_google_play`](https://github.com/janpio/fastlane-plugin-managed_google_play) and its actions `get_managed_play_store_publishing_rights` and `create_app_on_managed_play_store` into _fastlane_. 

### Description
This PR takes the currently existing functionality and adds it:

- `get_managed_play_store_publishing_rights` can be used to simplify the actication process of Managed Play Store publishing rights.
- `create_app_on_managed_play_store` can be used to then create a new app on a Managed Play Store.

This PR depends on https://github.com/fastlane/managed_google_play-callback having a working and published GitHub page version of the `callback.html` from the plugin.

I also created a repository that uses the modified version of fastlane and a plugin to use the new action to create apps online: https://github.com/janpio/fastlane-test-integrated-managed_google_play It replaces the `Fastfile` of the plugin that did the same job.

### Task List

- [x] Remove all TODOs from code and convert to items in this list
- [x] Move action description from README to action description in code
- [x] Find replacement for [testing `Fastfile`](https://github.com/janpio/fastlane-plugin-managed_google_play/blob/master/fastlane/Fastfile) (tests?)
- [x] Integrate new URL from published [fastlane/managed_google_play-callback](https://github.com/fastlane/managed_google_play-callback)
- [ ] Implementation TODOs
    - [x] Add some basic tests
    - [x] `json_key_data` as alternative to `json_key` to fill `credentials`
    - [x] Decide what to do with stolen `env_name` names (supply, produce)
    - [x] Investigate integration with `supply`
        - [ ] Use `supply` for API communication?
        - [ ] Use cleaned up `Supply::Options.available_options` instead of duplication?
        - [x] readd `root_url` and `timeout` parameters to `create_app_on_managed_play_store`?
        - [x] Move Google Play languages to `supply`?
    - [x] Add more tests
- [ ] Deprecate plugin

janpio committed:
Move over actions from plugin with minimal changes

janpio committed:
create empty spec files for actions

fastlane-bot-helper commented:
<!--
  0 Errors
  1 Warning: Pull Request is Work in Progre...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Pull Request is Work in Progress</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

janpio committed:
remove unneeded/broken require

janpio committed:
update author

janpio committed:
return URI in get_managedplay_store_publishing_rights

janpio committed:
make clear the details will be filled

janpio committed:
add action category

janpio committed:
make clear that URL will change

janpio committed:
example_code for get_managed_play_store_publishing_rights

janpio committed:
include documentation from README into details of get_managed_play_store_publishing_rights

janpio committed:
add description and example_code for create_app_on_managed_play_store

janpio committed:
better example

janpio committed:
fix code example

janpio committed:
remove root_url and timeout for now

janpio committed:
remove TODOs that were added to PR

janpio committed:
replace callback_uri with probable future URL

janpio committed:
rubocop

janpio committed:
rubocop 2

janpio committed:
Refactor Supply::Client to be usable by create_app_on_managed_play_store (#13431)

* fix order in verify_block for Supply's json_key:
check if file exists first, then if it is JSON

* Extract reusable parts of Client into GoogleServiceClient (self.make_from_config, initialize, call_google_api)

* Modify create_app_on_managed_play_store to implement and use Supply::GoogleServiceClient

* remove leftover code and TODOs
rubocop

* skip fastlane/spec/unused_options_spec.rb for new action

* remove TODO

* Rename Supply::GoogleServiceClient to Supply::AbstractGoogleServiceClient

* replce @scope/@service with SCOPE/SERVICE and use self.class::SCOPE/SERVICE to read it
https://stackoverflow.com/questions/19382626/how-do-i-override-a-variable-in-a-ruby-subclass-without-affecting-the-superclass/19382692#19382692

* workaround for `warning: already initialized constant Supply::UI`

* get rid of more code by moving client/android_publisher to AbstractGoogleServiceClient as client
(and also get rid of empty method implementations in PlaycustomappClient)

* move .new into initialize()

Commit:
Comment and environment variable name cleanup

Commit:
Moved some requires around

Commit:
Cleaned up some options on create_app_on_managed_play_store

Commit:
get_managed_play_store_publishing_rights works with json_key_data

Commit:
Rubo

Commit:
Added tests for get_managed_play_store_publishing_rights

Commit:
Rubo

Commit:
Removed unsured before in test

Commit:
Added some testing to create_app_on_managed_play_store

Commit:
Recommit to run tests maybe

Commit:
Moved languages into supply

jmarkoff reviewed (approved):
LGTM

Started from scratch and built a canonical Fastfile with a new app and a new Developer Account.

Thanks to @joshdholtz for the help.

Some notes to help others, make sure your release configuration has been setup to sign the APK. You can use something like this to do so.

```
ENV['SUPPLY_JSON_KEY'] = 'key.json'
ENV['SUPPLY_DEVELOPER_ACCOUNT_ID'] = '111111111111000000000'
ENV['SUPPLY_APP_TITLE'] = 'APP TITLE'
// Added the following env vars to ~/.gradle/gradle.properties
signingConfigs {
        release {
            storeFile file(RELEASE_STORE_FILE)
            storePassword RELEASE_STORE_PASSWORD
            keyAlias RELEASE_KEY_ALIAS
            keyPassword RELEASE_KEY_PASSWORD
        }
    }

buildTypes {
        release {
            signingConfig signingConfigs.release
       }
}

```
Fastfile:

```
platform :android do

  desc "Enable Private App upload for your domain on Managed Google Play"
  lane :enable_private_apps do
    get_managed_play_store_publishing_rights(
        json_key: ENV['SUPPLY_JSON_KEY']
    )
  end

  desc "Create the private app on the Google Play store"
  lane :create_private_app do
      gradle(
        task: 'assemble',
        build_type: 'Release'
      )

      # Finds latest APK
      apk_path = Actions.lane_context[SharedValues::GRADLE_APK_OUTPUT_PATH]

      create_app_on_managed_play_store(
        json_key: ENV['SUPPLY_JSON_KEY'],
        developer_account_id: ENV['SUPPLY_DEVELOPER_ACCOUNT_ID'],
        app_title: ENV['SUPPLY_APP_TITLE'],
        language: "en_US",
        apk: apk_path
      )
    end
end
```

Commit:
Remove commented code

janpio commented on fastlane/lib/fastlane/actions/create_app_on_managed_play_store.rb:
```diff
@@ -85,24 +85,18 @@ def self.available_options
             short_option: "-k",
             env_name: "SUPPLY_DEVELOPER_ACCOUNT_ID",
             description: "The ID of your Google Play Console account. Can be obtained from the URL when you log in (`https://play.google.com/apps/publish/?account=...` or when you 'Obtain private app publishing rights' (https://developers.google.com/android/work/play/custom-app-api/get-started#retrieve_the_developer_account_id)",
-            optional: false,
```
Is `optional: false` the default or why could these (this one and the others from the params) be removed?

janpio commented:
Thanks @joshdholtz for the further cleanup and tests - looking good!
